src/Controller/Api/Legacy/GalleryController.php

```diff
@@ -0,0 +1,520 @@
+     */
+    protected function handleWriteGalleryhasmedia(GalleryInterface $gallery, MediaInterface $media, ?GalleryHasMediaInterface $galleryHasMedia = null, Request $request)
+    {
+        $form = $this->formFactory->createNamed(null, ApiGalleryHasMediaType::class, $galleryHasMedia, [
+            'csrf_protection' => false,
+        ]);
+
+        $form->handleRequest($request);
+
+        if ($form->isValid()) {
+            $galleryHasMedia = $form->getData();
+            $galleryHasMedia->setMedia($media);
+
+            // NEXT_MAJOR: remove this if/else block. Just call `$gallery->addGalleryHasMedia($galleryHasMedia);`
+            if ($gallery instanceof GalleryMediaCollectionInterface) {
+                $gallery->addGalleryHasMedia($galleryHasMedia);
+            } else {
+                $gallery->addGalleryHasMedias($galleryHasMedia);
+            }
+            $this->galleryManager->save($gallery);
+
+            $context = new Context();
+            $context->setGroups(['sonata_api_read']);
+            $context->enableMaxDepth();
+
+            $view = FOSRestView::create($galleryHasMedia);
+            $view->setContext($context);
+
+            return $view;
+        }
+
+        return $form;
+    }
+
+    /**
+     * Retrieves gallery with identifier $id or throws an exception if it doesn't exist.
+     *
+     * @param string $id Gallery identifier
+     *
+     * @throws NotFoundHttpException
+     *
+     * @return GalleryInterface
+     */
+    protected function getGallery($id)
+    {
+        $gallery = $this->getGalleryManager()->findOneBy(['id' => $id]);
+
+        if (null === $gallery) {
+            throw new NotFoundHttpException(sprintf('Gallery (%s) not found', $id));
+        }
+
+        return $gallery;
+    }
+
+    /**
+     * Retrieves media with identifier $id or throws an exception if it doesn't exist.
+     *
+     * @param string $id Media identifier
+     *
+     * @throws NotFoundHttpException
+     *
+     * @return MediaInterface
+     */
+    protected function getMedia($id)
+    {
+        $media = $this->getMediaManager()->findOneBy(['id' => $id]);
+
+        if (null === $media) {
+            throw new NotFoundHttpException(sprintf('Media (%s) not found', $id));
+        }
+
+        return $media;
+    }
+
+    /**
+     * @return GalleryManagerInterface
+     */
+    protected function getGalleryManager()
+    {
+        return $this->galleryManager;
+    }
+
+    /**
+     * @return MediaManagerInterface
+     */
+    protected function getMediaManager()
+    {
+        return $this->mediaManager;
+    }
+
+    /**
+     * Write a Gallery, this method is used by both POST and PUT action methods.
+     *
+     * @param Request     $request Symfony request
+     * @param string|null $id      Gallery identifier
+     *
+     * @return Rest\View|FormInterface
+     */
+    protected function handleWriteGallery($request, $id = null)
+    {
+        $gallery = $this->getGallery($id);
```

Comment:
If we are allowing empty values (`''`, `null`) to be used as argument 1 for `getGallery()`, I think we should improve the exception message thrown there. Otherwise it will end as:
>Gallery () not found

Reply:
Gallery with identifier "" is not found ??

Reply:
I agree :+1: 
IMO, using quotes instead of parentheses as value delimiter seems more understandable.

Reply:
@phansys WDYT about?:
```php
throw new NotFoundHttpException(sprintf(
            'Media %s not found',
            empty($id) ? ('' === $id ? 'for empty string identifier is' : 'for null identifier is') : '('.$id.')'
        ));
```

Reply:
IMO, using `""` for these identifiers is enough, but in case you get traction with this suggestion, please note the following:
```php
empty('0') === true
```

Reply:
What about?:
```
throw new NotFoundHttpException(sprintf('Media not found - identifier: %s.', print_r($id, true)));
```

Reply:
```
throw new NotFoundHttpException(sprintf('Media not found - identifier: %s.', var_export($id, true)));
```
For null: Gallery not found - identifier: NULL
For '42': Gallery not found - identifier: '42'
For 'string': Gallery not found - identifier: 'string'
For '': Gallery not found - identifier: ''
For 42:  Gallery not found - identifier: 42

Reply:
I think we could use this approach, using a colloquial syntax. I mean something like the following example or similar:
```php
sprintf('Media not found for identifier %s.', var_export($id, true));
```
Regardless that, I agree to use `var_export()`.

Reply:
Done.